packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -1742,46 +1440,599 @@ class RenderViewport2 extends RenderBox with ContainerRenderObjectMixin<RenderSl
+    assert(extent >= 0.0);
+    assert(crossExtent.isFinite);
+    assert(crossExtent >= 0.0);
+    assert(correctedOffset.isFinite);
+    _maxScrollExtent = 0.0;
+    _shrinkWrapExtent = 0.0;
+    _hasVisualOverflow = false;
+    return layoutOneSide(
+      firstChild,
+      math.max(0.0, correctedOffset),
+      0.0,
+      extent,
+      crossExtent,
+      GrowthDirection.forward,
+      childAfter,
+    );
+  }
+
+  @override
+  bool get hasVisualOverflow => _hasVisualOverflow;
+
+  @override
+  void updateOutOfBoundsData(GrowthDirection growthDirection, SliverGeometry childLayoutGeometry) {
+    assert(growthDirection == GrowthDirection.forward);
+    _maxScrollExtent += childLayoutGeometry.scrollExtent;
+    if (childLayoutGeometry.hasVisualOverflow)
+      _hasVisualOverflow = true;
+    _shrinkWrapExtent += childLayoutGeometry.maxPaintExtent;
+  }
+
+  @override
+  void updateChildPaintOffset(RenderSliver child, double paintOffset, GrowthDirection growthDirection) {
+    assert(growthDirection == GrowthDirection.forward);
+    final SliverLogicalParentData childParentData = child.parentData;
+    childParentData.scrollOffset = paintOffset;
+  }
+
+  @override
+  Offset paintOffsetOf(RenderSliver child) {
+    final SliverLogicalParentData childParentData = child.parentData;
+    return computeAbsolutePaintOffset(child, childParentData.scrollOffset, GrowthDirection.forward);
+  }
+
+  @override
+  void applyPaintTransform(RenderObject child, Matrix4 transform) {
+    assert(child != null);
+    Offset offset = paintOffsetOf(child);
+    transform.translate(offset.dx, offset.dy);
+  }
+
+  @override
+  double computeChildMainAxisPosition(RenderSliver child, double parentMainAxisPosition) {
+    assert(child != null);
+    assert(child.constraints != null);
+    assert(hasSize);
+    final SliverLogicalParentData childParentData = child.parentData;
+    switch (applyGrowthDirectionToAxisDirection(child.constraints.axisDirection, child.constraints.growthDirection)) {
+      case AxisDirection.up:
+        return (size.height - parentMainAxisPosition) - childParentData.scrollOffset;
+      case AxisDirection.right:
+      case AxisDirection.down:
+        return parentMainAxisPosition - childParentData.scrollOffset;
+      case AxisDirection.left:
+        return (size.width - parentMainAxisPosition) - childParentData.scrollOffset;
+    }
+    return 0.0;
+  }
+
+  @override
+  void rereportDimensions() {
+    assert(_maxScrollExtent != null);
+    double effectiveExtent;
+    switch (axis) {
+      case Axis.vertical:
+        effectiveExtent = size.height;
+        break;
+      case Axis.horizontal:
+        effectiveExtent = size.width;
+        break;
+    }
+    assert(effectiveExtent != null);
+    offset.applyViewportDimension(effectiveExtent);
+    if (!offset.applyContentDimensions(0.0, math.max(0.0, _maxScrollExtent - effectiveExtent)))
+      markNeedsLayout();
+  }
+
+  @override
+  int indexOfFirstChild() => 0;
```

Comment:
because one of its implementation is O(N) and our design guide says to use methods for expensive getters.